src/Azure/Orleans.ScheduledJobs.AzureStorage/AzureStorageJobShard.cs

```diff
@@ -124,16 +135,71 @@ public async ValueTask InitializeAsync(CancellationToken cancellationToken)
         ETag = response.Value.Details.ETag;
     }
 
-    private async Task AppendOperation(JobOperation operation, CancellationToken cancellationToken)
+    private async Task EnqueueStorageOperationAsync(StorageOperation operation, CancellationToken cancellationToken)
+    {
+        var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        operation.CompletionSource = tcs;
+        
+        await _storageOperationChannel.Writer.WriteAsync(operation, cancellationToken);
+        await tcs.Task;
+    }
+
+    private async Task ProcessStorageOperationsAsync()
+    {
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ContinueOnCapturedContext | ConfigureAwaitOptions.ForceYielding);
+
+        var cancellationToken = _shutdownCts.Token;
+        
+        try
+        {
+            await foreach (var operation in _storageOperationChannel.Reader.ReadAllAsync(cancellationToken))
+            {
+                try
+                {
+                    switch (operation.Type)
+                    {
+                        case StorageOperationType.AppendJobOperation:
+                            await AppendJobOperationAsync(operation.JobOperation!.Value, cancellationToken);
+                            break;
+                        case StorageOperationType.UpdateMetadata:
+                            await UpdateMetadataAsync(operation.Metadata!, cancellationToken);
+                            break;
+                    }
+                    
+                    operation.CompletionSource?.TrySetResult(true);
+                }
+                catch (Exception ex)
+                {
+                    operation.CompletionSource?.TrySetException(ex);
+                }
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected during shutdown
+        }
+    }
```

Comment:
I think we need to cancel all pending operations after completing the channel and exiting this loop. Otherwise, there could be a hang